beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.beam.direct.io;
+
+import com.google.common.base.Preconditions;
+import cz.o2.proxima.direct.commitlog.Offset;
+import java.io.Serializable;
+import java.util.Objects;
+import lombok.Getter;
+import lombok.ToString;
+import org.apache.beam.sdk.transforms.splittabledofn.HasDefaultTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.transforms.splittabledofn.SplitResult;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * A {@link RestrictionTracker} for {@link Offset Offsets} read from {@link
+ * cz.o2.proxima.direct.commitlog.CommitLogReader}
+ */
+public class OffsetRestrictionTracker
+    extends RestrictionTracker<OffsetRestrictionTracker.StartEndOffset, Offset> {
+
+  @ToString
+  public static class StartEndOffset
```

Comment:
why do we need this to be Serializable?

Reply:
It uses SerializableCoder, and there is no other way to serialize the `Offset`s.

Reply:
It may be tricky, but I think we should get rid of this at the beginning as this caused so much trouble in the past.

- We can introduce a new interface `ExternalizableOffset extends Offset`, which would introduce `encode` / `decode` methods. This way we can incrementally add support for different offset implementations.
- Once we have a clean way to serialize offsets, we can simply write `StartEndOffset` coder.

Reply:
I'll think about that. Definitely it is possible.

Reply:
O2-Czech-Republic#181